include/deal.II/lac/tensor_product_matrix.h

```diff
@@ -311,96 +691,193 @@ template <int dim, typename Number, int size>
+  auto &&derivative_matrix = std::forward<MatrixArray>(derivative_matrices_) ;
+  this->mass_matrix = mass_matrix ;
+  this->derivative_matrix = derivative_matrix ;
+
+  constexpr unsigned int macro_size = VectorizedArray<Number>::n_array_elements ;
+  const unsigned int nm_flat_size
+    = (size > 0)
+      ? (Utilities::fixed_int_power<size,dim>::value
+         * Utilities::fixed_int_power<size,dim>::value * macro_size)
+      : (Utilities::fixed_power<dim>(mass_matrix[0].n_rows())
+         * Utilities::fixed_power<dim>(mass_matrix[0].n_rows()) * macro_size) ;
+  const unsigned int n_flat_size
+    = (size > 0)
+      ? Utilities::fixed_int_power<size,dim>::value * macro_size
+      : Utilities::fixed_power<dim>(mass_matrix[0].n_rows()) * macro_size ;
+
+  std::vector<Number> mass_matrix_flat ;
+  std::vector<Number> deriv_matrix_flat ;
+  std::vector<Number> eigenvalues_flat ;
+  std::vector<Number> eigenvectors_flat ;
+  mass_matrix_flat.reserve (nm_flat_size) ;
+  deriv_matrix_flat.reserve (nm_flat_size) ;
+  eigenvalues_flat.reserve (n_flat_size) ;
+  eigenvectors_flat.reserve (nm_flat_size) ;
+  std::array<unsigned int,macro_size> offsets_nm ;
+  std::array<unsigned int,macro_size> offsets_n ;
+  for (int dir = 0; dir < dim; ++dir)
     {
-      eval.template apply<0, true, false>(M, src, t+n);
-      eval.template apply<1, true, false>(M, t+n, t);
-      eval.template apply<2, true, false>(A, t, dst);
-      eval.template apply<1, true, false>(A, t+n, t);
-      eval.template apply<0, true, false>(A, src, t+n);
-      eval.template apply<1, true, true> (M, t+n, t);
-      eval.template apply<2, true, true> (M, t, dst);
+      Assert (size == -1 ||
+              (size > 0 && static_cast<unsigned int>(size) == mass_matrix[dir].n_rows()),
+              ExcDimensionMismatch(size, mass_matrix[dir].n_rows()));
+      AssertDimension (mass_matrix[dir].n_rows(), mass_matrix[dir].n_cols());
+      AssertDimension (mass_matrix[dir].n_rows(), derivative_matrix[dir].n_rows());
+      AssertDimension (mass_matrix[dir].n_rows(), derivative_matrix[dir].n_cols());
+
+      const unsigned int n_rows = mass_matrix[dir].n_rows() ;
+      const unsigned int n_cols = mass_matrix[dir].n_cols() ;
+      const unsigned int nm = n_rows * n_cols ;
+
+      mass_matrix_flat.resize (macro_size*nm) ;
+      deriv_matrix_flat.resize (macro_size*nm) ;
+      eigenvalues_flat.resize (macro_size*n_rows) ;
+      eigenvectors_flat.resize (macro_size*nm) ;
+      for (unsigned int vv=0; vv<macro_size; ++vv)
+        offsets_nm[vv] = nm * vv ;
+      for (unsigned int vv=0; vv<macro_size; ++vv)
+        offsets_n[vv] = n_rows * vv ;
+
+      vectorized_transpose_and_store (false, nm, &(mass_matrix[dir](0,0))
+                                      , offsets_nm.cbegin(), mass_matrix_flat.data()) ;
+      vectorized_transpose_and_store (false, nm, &(derivative_matrix[dir](0,0))
+                                      , offsets_nm.cbegin(), deriv_matrix_flat.data()) ;
```

Comment:
That's neat, I hadn't thought about using it here, but it definitely makes sense.